lib/std/debug.zig

```diff
@@ -291,43 +285,61 @@ pub fn writeStackTrace(
 }
 
 pub const StackIterator = struct {
-    first_addr: ?usize,
+    // Skip every frame before this address is found
+    first_address: ?usize,
+    // Last known value of the frame pointer register
     fp: usize,
 
-    pub fn init(first_addr: ?usize) StackIterator {
+    pub fn init(first_address: ?usize, fp: ?usize) StackIterator {
         return StackIterator{
-            .first_addr = first_addr,
-            .fp = @frameAddress(),
+            .first_address = first_address,
+            .fp = fp orelse @frameAddress(),
         };
     }
 
     // On some architectures such as x86 the frame pointer is the address where
     // the previous fp is stored, while on some other architectures such as
     // RISC-V it points to the "top" of the frame, just above where the previous
     // fp and the return address are stored.
-    const fp_adjust_factor = if (builtin.arch == .riscv32 or builtin.arch == .riscv64)
+    const fp_offset = if (builtin.arch.isRISCV())
         2 * @sizeOf(usize)
     else
         0;
 
     fn next(self: *StackIterator) ?usize {
-        if (self.fp <= fp_adjust_factor) return null;
-        self.fp = @intToPtr(*const usize, self.fp - fp_adjust_factor).*;
-        if (self.fp <= fp_adjust_factor) return null;
-
-        if (self.first_addr) |addr| {
-            while (self.fp > fp_adjust_factor) : (self.fp = @intToPtr(*const usize, self.fp - fp_adjust_factor).*) {
-                const return_address = @intToPtr(*const usize, self.fp - fp_adjust_factor + @sizeOf(usize)).*;
-                if (addr == return_address) {
-                    self.first_addr = null;
-                    return return_address;
-                }
+        var address = self.next_internal() orelse return null;
+
+        if (self.first_address) |first_address| {
+            while (address != first_address) {
+                address = self.next_internal() orelse return null;
             }
+            self.first_address = null;
         }
 
-        const return_address = @intToPtr(*const usize, self.fp - fp_adjust_factor + @sizeOf(usize)).*;
-        if (return_address == 0) return null;
-        return return_address;
+        return address;
+    }
+
+    fn next_internal(self: *StackIterator) ?usize {
+        const fp = math.sub(usize, self.fp, fp_offset) catch return null;
+
+        // Sanity check
+        if (fp == 0 or !mem.isAligned(fp, @alignOf(usize)))
+            return null;
+
+        const new_fp = @intToPtr(*const usize, fp).*;
+
+        // Sanity check: the stack grows down thus all the parent frames must be
+        // be at addresses that are greater (or equal) than the previous one.
+        // A zero frame pointer often signals this is the last frame, that case
+        // is gracefully handled by the next call to next_internal
+        if (new_fp != 0 and new_fp < self.fp)
+            return null;
```

Comment:
What about `sigaltstack` and `@call(.{.stack = ...}, ...)`?

Reply:
That's a good point - I will revert this part unless @LemonBoy wants to make a case to keep it.

Reply:
It was a deliberate design decision, frame walking is not that robust by itself so we have to try hard not to fuck up too bad.. you don't want misleading stack traces or worse no trace at all.

> What about sigaltstack

The signal handler we use doesn't care as it pulls the `ip` and `fp` out of a structure and, at the time of writing, the handler doesn't take advantage of `sigaltstack` at all.
Pulling this trick from a signal handler is extremely tricky (signals are extremely tricky in general) and the old code doesn't work either in such a situation, see the code under the spoiler:

<details>

```zig
const std = @import("std");
const os = std.os;

fn handler(sig: i32, info: *const os.siginfo_t, ctx_ptr: *const c_void) callconv(.C) noreturn {
    @panic("segfault");
}

pub fn main() !void {
    var stack = try os.mmap(
        null,
        os.MINSIGSTKSZ * 10,
        os.PROT_WRITE | os.PROT_READ,
        os.MAP_PRIVATE | os.MAP_ANONYMOUS,
        -1,
        0,
    );
    var st = os.stack_t{
        .ss_sp = stack.ptr,
        .ss_flags = 0,
        .ss_size = stack.len,
    };
    try os.sigaltstack(&st, null);

    var act = os.Sigaction{
        .sigaction = handler,
        .mask = os.empty_sigset,
        .flags = (os.SA_SIGINFO | os.SA_RESTART | os.SA_RESETHAND | os.SA_ONSTACK),
    };
    os.sigaction(os.SIGSEGV, &act, null);
    os.sigaction(os.SIGILL, &act, null);

    @intToPtr(*allowzero usize, 0).* = 1;
}
```

</details>

> and @call(.{.stack = ...}, ...)?

Since it's not really working and is a bit of a niche operation (that IMO doesn't deserve a place in `@call`) I couldn't care less about that, I'm more interested in getting good traces for the 99.99% remaining cases.

Reply:
Thanks for the explanation. I'll put this explanation into comments in the code

Reply:
Will this result in aborting at the first differed stack? That is, *if* you use `@call(.{stack = ...`, will the stack trace end at the function that calls? Or will it fail entirely?

Reply:
In that case the behaviour is undefined. It may or may not print the stack trace before the call on a new stack, it depends on the location of that stack in virtual memory.